File: kotlin/src/main/kotlin/com/github/jntakpe/aoc2021/days/day7/Day7.kt
package com.github.jntakpe.aoc2021.days.day7

import com.github.jntakpe.aoc2021.shared.Day
import com.github.jntakpe.aoc2021.shared.readInputLines
import kotlin.math.abs

object Day7 : Day {

    override val input = readInputLines(7).joinToString("").split(",").map { it.toInt() }

    override fun part1() = resolve { from, to -> abs(from - to) }

    override fun part2() = resolve { from, to -> (0..abs(from - to)).sum() }

    private fun resolve(fuel: (Int, Int) -> Int) = (input.minOf { it }..input.maxOf { it }).minOf { d -> input.sumOf { fuel(it, d) } }
}

File: rust/src/bin/day7.rs
use aoc_2021::shared::{Day, read_input_lines};

fn main() {
    Day7::new(read_input_lines(7)).run()
}

struct Day7 {
    numbers: Vec<usize>,
}

impl Day for Day7 {
    fn part1(&self) -> usize {
        self.resolve(gap)
    }

    fn part2(&self) -> usize {
        self.resolve(|from, dest| (0..=gap(from, dest)).sum())
    }
}

impl Day7 {
    fn new(input: Vec<String>) -> Day7 {
        Day7 { numbers: input.join("").split(',').map(|s| s.parse::<usize>().unwrap()).collect() }
    }

    fn resolve(&self, fuel: fn(usize, usize) -> usize) -> usize {
        let min = *self.numbers.iter().min().unwrap();
        let max = *self.numbers.iter().max().unwrap();
        (min..max).map(|dest| self.numbers.iter().map(|from| fuel(*from, dest)).sum()).min().unwrap()
    }
}

fn gap(from: usize, dest: usize) -> usize {
    (from as isize - dest as isize).abs() as usize
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part1_sample() {
        assert_eq!(sample_day().part1(), 37)
    }

    #[test]
    fn part2_sample() {
        assert_eq!(sample_day().part2(), 168)
    }

    fn sample_day() -> Day7 {
        Day7::new(vec![String::from("16,1,2,0,4,2,7,1,2,14")])
    }
}
